### PR TITLE
[do not merge] 512 multipart 

### PR DESCRIPTION
**Pull request by gentee**

### CODE REVIEW

File: packages/api/contract.go [hunk 1]
```diff
@@ -92,6 +94,35 @@ func contract(w http.ResponseWriter, r *http.Request, data *apiData, logger *log
...
+					if _, fileHeader, _ := r.FormFile(fitem.Name); fileHeader != nil {
+						file, err := fileHeader.Open()
+						if err != nil {
+							log.WithFields(log.Fields{"type": consts.InvalidObject, "error": err}).Error("getting multipart file")
+							return errorAPI(w, err.Error(), http.StatusBadRequest)
+						}
+						buf, err := ioutil.ReadAll(file)
+						file.Close()
+						if err != nil {
+							log.WithFields(log.Fields{"type": consts.InvalidObject, "error": err}).Error("reading multipart file")
```
Comment (potehinre) [inline packages/api/contract.go]: type: consts.IOError

File: packages/api/data.go [hunk 2]
```diff
@@ -62,8 +63,14 @@ func dataHandler() hr.Handle {
 			errorAPI(w, `E_NOTFOUND`, http.StatusNotFound)
 			return
 		}
+		if mime := smart.FileMime([]byte(data)); len(mime) > 0 {
+			w.Header().Set("Content-Type", mime)
+			w.Header().Set("Cache-Control", "public,max-age=604800,immutable")
```
Comment (potehinre) [inline packages/api/data.go]: move max_age to consts

File: packages/api/prepare.go [hunk 3]
```diff
@@ -62,13 +66,39 @@ func prepareContract(w http.ResponseWriter, r *http.Request, data *apiData, logg
...
+			if strings.Contains(fitem.Tags, `signature`) {
 				continue
 			}
 			var val string
-			if fitem.Type.String() == `[]interface {}` {
+			if strings.Contains(fitem.Tags, `file`) && data.multipart {
+				if _, fileHeader, _ := r.FormFile(fitem.Name); fileHeader != nil {
+					file, err := fileHeader.Open()
+					if err != nil {
+						log.WithFields(log.Fields{"type": consts.InvalidObject, "error": err}).Error("getting multipart file")
```
Comment (potehinre) [inline packages/api/prepare.go]: type: consts.IOError

File: packages/api/prepare.go [hunk 4]
```diff
@@ -62,13 +66,39 @@ func prepareContract(w http.ResponseWriter, r *http.Request, data *apiData, logg
...
+						return errorAPI(w, err.Error(), http.StatusBadRequest)
+					}
+					buf, err := ioutil.ReadAll(file)
+					file.Close()
+					if err != nil {
+						log.WithFields(log.Fields{"type": consts.InvalidObject, "error": err}).Error("reading multipart file")
```
Comment (potehinre) [inline packages/api/prepare.go]: type: consts.IOError

File: packages/parser/common_parse_data_full.go [hunk 5]
```diff
@@ -359,13 +359,52 @@ func parseContractTransaction(p *Parser, buf *bytes.Buffer) error {
 
 	input := smartTx.Data
 	p.TxData = make(map[string]interface{})
-
+	getHash := func(input []byte) (string, error) {
```
Comment (potehinre) [inline packages/parser/common_parse_data_full.go]: Code duplication, move that func to utils

**Review by potehinre — CHANGES_REQUESTED** [review]

File: packages/api/api.go [hunk 6]
```diff
@@ -221,7 +222,12 @@ func DefaultHandler(method, pattern string, params map[string]int, handlers ...a
 			}
 		}
 		// Getting and validating request parameters
-		r.ParseForm()
+		data.multipart = strings.HasPrefix(r.Header.Get(`Content-Type`), `multipart`)
+		if data.multipart {
+			r.ParseMultipartForm(15 * 1024 * 1024)
```
[the remark this answers is not in the document]
Comment (gentee) [reply packages/api/api.go]: Fixed all review comments

**Review by potehinre — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request closed without merging**